### DISCUSSION TITLE
Proper way to develop with babel in ESM projects of type module

### DISCUSSION THREAD

**Discussion by dlong500** [babel/babel#13579] [Q&A]
What is the currently recommended alternative to @babel/register in ESM type module projects?  I understand that babel is compatible with ESM, but when developing how can on-demand transforming be achieved?

I don't actually care so much about transpilation of code features (though that is still a valid use case with next-gen ES features not supported by the current version of Node), but how can things like **module resolution (provided by babel-plugin-module-resolver)** be applied in a development environment when it seems like @babel/register and @babel/node don't support ESM contexts? (See #11108).

**Comment by nicolo-ribaudo** [comment]
Node.js doesn't provide a stable loaders hook API for ESM yet, so there isn't any official package for that. We suggest compiling your code using `@babel/cli`, and then running the compiled code.

**Reply by dlong500** [reply]
Any general idea when Node.js is going to finalize their ESM loaders hook?  I can't seem to find any mention of a timeline or specific work on the feature.  It's a shame because a bunch of package maintainers have decided to go all in with pure ESM (with no CJS fallback).  So now the build tooling is falling behind a bit because of things like the Node issue you mentioned.

**Reply by nicolo-ribaudo** [reply]
All the discussion is happening at https://github.com/nodejs/modules/issues/351 and https://github.com/nodejs/loaders.

**Reply by dlong500** [reply]
I really appreciate the links and feedback.  Although I'm not sure what to make of the fact that there hasn't even been a post in that nodejs/modules issue in over a year.  I hope this can get some traction sooner rather than later.

**Reply by nicolo-ribaudo** [reply]
I think that the issue hasn't been updated because the discussion has been moved to the new repo.

If you are interested, this repository has a Node.js loader for Babel (but since the Node.js API is experimental and it will change, it's not production ready): https://github.com/targos/multiloader
